checks for samples not at time 0

petrelharp opened the pull request:
I've added some tests; they look good.

I also removed the requirement that get_time() apply only to samples, as it was previously checking by comparing to num_samples - I can't see what this would break and the tests pass...

jeromekelleher commented on msprime/trees.py:
```diff
@@ -2320,7 +2320,6 @@ def node(self, u):
         return self.get_time(sample)
 
-    # TODO refactor these methods to take node arguments not samples.
     def get_time(self, sample):
```
We should change the parameter here to ``u`` (also for the ``time`` method above), and change the documentation so that sample -> node.

petrelharp replied:
it is done.  also for get_population.

jeromekelleher commented:
Hah, it's pretty cool that worked without any problems. So, there's no reason at all that time has to start at 0. It should also work if we have negative time I guess. Is there any point in allowing this?

petrelharp commented:
Hey, that's great.  Yes - then switching from forwards to reverse time
would just involve multiplying by -1, rather than having to go look up the
maximum time and subtract that.

On Fri, Apr 28, 2017 at 1:37 AM, Jerome Kelleher <notifications@github.com>
wrote:

> Hah, it's pretty cool that worked without any problems. So, there's no
> reason at all that time has to start at 0. It should also work if we have
> negative time I guess. Is there any point in allowing this?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/jeromekelleher/msprime/pull/199#issuecomment-297941598>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA_26VVeY37fkl-BZvtSVcTvPe9xb3Erks5r0aU3gaJpZM4NK4Rj>
> .
>

petrelharp commented:
OK - this is ready to merge (pending what @jeromekelleher wants to do about new check for internal samples).

ashander commented:
@petrelharp there were a couple issues breaking c-level tests. I was going to add comments here on the diff but wanted to make sure they'd work, so just pushed a PR to your `generalise-samples` branch (see above).

petrelharp commented:
Darn it, forgot to run those check, thanks.  I've checked and merged.

But - at first I couldn't check because compiling the tests gave me this error:
```
msprime/lib$ make
(...)
tree_sequence.c: In function ‘tree_sequence_write_hdf5_metadata’:
tree_sequence.c:2148:22: error: conversion to ‘herr_t {aka int}’ from ‘hid_t {aka long int}’ may alter its value [-Werror=conversion]
             status = dataspace_id;
```

Casting `dataspace_id` to `herr_t` makes this complie; I've done this but don't actually know if this might have bad effects. (@jeromekelleher)

ashander commented:
As a data point: I didn't run into that `herr_t` vs `hid_t` error/warning.

Also @petrelharp I think you may have forgotten to push your merge. In any case, since you gave me access to your fork, I just merged my commit to set if it'll work here on CI

jeromekelleher commented:
Thanks a million @petrelharp and @ashander, this is great. I've made a few minor changes:

- Move the test for internal nodes as samples to ``init_nodes``. This is just to try and keep all the error checking logic grouped together in some way. Currently there are two types of error checking 'cheap' error checks that we do everytime we load a tree sequence and more expensive checks that are done in ``tree_sequence_check``. The distinction between the two is somewhat arbitrary at the moment...

- Refactored the tests in ``test_topology`` a little bit.

Re the ``herr_t`` vs ``hid_t``  thing, this is because @petrelharp is using a different compiler. Casting should be harmless and is the right thing to do. Is the change included in this patch?

Unfortunately, I didn't spot that this PR was against the ``generalise-samples`` branch until after I had merged it. However, if you rebase this PR off master we should be able to merge it easily then.

ashander commented:
@jeromekelleher @petrelharp I rebased it and add that change (see https://github.com/jeromekelleher/msprime/pull/199/commits/2ef42a26b3896be8e2e3f5da6067c7849f8c5e4d ) which was there, but just on `petrelharp/master`

Note: I can't change the base of this to `master` so it looks like some of Jerome's commits from `generalise-samples` are still included here

petrelharp commented:
So that's where that commit went.  Whoops!  Thanks, Jaime.  I've also
changed the base.

On Tue, May 2, 2017 at 9:46 AM, ashander <notifications@github.com> wrote:

> @jeromekelleher <https://github.com/jeromekelleher> @petrelharp
> <https://github.com/petrelharp> I rebased it and add that change (see
> 2ef42a2
> <https://github.com/jeromekelleher/msprime/pull/199/commits/2ef42a26b3896be8e2e3f5da6067c7849f8c5e4d>
> ) which was there, but just on petrelharp/master
>
> Note: I can't change the base of this to master so it looks like some of
> Jerome's commits from generalise-samples are still included here
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/jeromekelleher/msprime/pull/199#issuecomment-298692238>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA_26bs6YkUqkWToFxF4MVaaPrxxrYHbks5r113JgaJpZM4NK4Rj>
> .
>

jeromekelleher commented:
Merging; thanks @petrelharp and @ashander!
